pkg/virt-handler/rest/console.go

```diff
@@ -0,0 +1,195 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2017 Red Hat, Inc.
+ *
+ */
+
+package rest
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"path"
+	"strconv"
+	"sync"
+
+	"github.com/emicklei/go-restful"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+
+	v1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+	"kubevirt.io/kubevirt/pkg/virt-handler/isolation"
+)
+
+type ConsoleHandler struct {
+	podIsolationDetector isolation.PodIsolationDetector
+	serialStopChans      map[types.UID](chan struct{})
+	vncStopChans         map[types.UID](chan struct{})
+	serialLock           *sync.Mutex
+	vncLock              *sync.Mutex
+	vmiInformer          cache.SharedIndexInformer
+}
+
+func NewConsoleHandler(podIsolationDetector isolation.PodIsolationDetector, vmiInformer cache.SharedIndexInformer) *ConsoleHandler {
+	return &ConsoleHandler{
+		podIsolationDetector: podIsolationDetector,
+		serialStopChans:      make(map[types.UID](chan struct{})),
+		vncStopChans:         make(map[types.UID](chan struct{})),
+		serialLock:           &sync.Mutex{},
+		vncLock:              &sync.Mutex{},
+		vmiInformer:          vmiInformer,
+	}
+}
+
+func (t *ConsoleHandler) VNCHandler(request *restful.Request, response *restful.Response) {
+	vmi, err := t.getVMI(request)
+	if err != nil {
+		log.Log.Reason(err).Error("Failed retrieve VMI")
+		response.WriteError(http.StatusBadRequest, err)
+		return
+	}
+	unixSocketPath, err := t.getUnixSocketPath(vmi, "virt-vnc")
```

Comment:
You will have to check if the path exists. If it does not exist, consult the `domain` cache and see if the VMI was reported as down, or as not existent, to return appropriate errors.

Reply:
can you elaborate -
1. by `domain` you mean libvirt domain? or you mean to query the vmi from vmiinformer?
2. you mean when `Dial`ing the socket fails?

Reply:
Hm, a little bit tricky to do the right thing here. Le'ts keep it as you have it (I meant querying the domain-cache, similar to the vmi domain, just for libvirt domains, the virt-handler has that too).

Reply:
I think on the long run we will do a "consolidate" call and directly check the domain state to see why it could not connect and return an appropriate error.

Reply:
ok